onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc

```diff
@@ -831,23 +831,74 @@ onnxruntime::Node& NodeFromApiNode(onnx_layout_transformation::api::NodeRef& nod
   return static_cast<ApiNode&>(node).Node();
 }
 
+CostCheckResult OrtDefaultCostCheck(const api::GraphRef& graph, const api::NodeRef& node,
+                                    const std::vector<int64_t>& /*perm*/,
+                                    const std::unordered_set<std::string>& /*outputs_leading_to_transpose*/) {
+  // special case some kernels based on the ORT implementation details
+  if (node.GetExecutionProviderType() == kCpuExecutionProvider) {
+    if (node.IsOp("MaxPool")) {
+      // MaxPool has higher perf in the NHWC variant when supported. HandleMaxPool does the support checks.
+      return CostCheckResult::kPushTranspose;
+    }
+
+    if (node.IsOp("Resize")) {
+      // Resize is included because it has higher perf in the NHWC variant when
+      // the input X is 4D int8 tensor and the mode is linear
+      auto X_value_info = graph.GetValueInfo(node.Inputs()[0]);
+      auto X_shape = X_value_info->Shape();
+      auto X_dtype = X_value_info->DType();
+      auto mode = node.GetAttributeString("mode");
+      if (X_shape && X_shape->size() == 4 &&
+          (X_dtype == api::DataType::UINT8 || X_dtype == api::DataType::INT8) &&
+          mode && *mode == "linear") {
+        return CostCheckResult::kPushTranspose;
+      }
+    }
+  }
+
+  return CostCheckResult::kFallThrough;
+}
+
 namespace layout_transformer {
 
 const std::unordered_set<std::string_view>& GetORTLayoutSensitiveOps() {
   static std::unordered_set<std::string_view> ort_layout_sensitive_ops = []() {
     const auto& layout_sensitive_ops = onnx_layout_transformation::GetLayoutSensitiveOps();
-#if !defined(USE_CUDA) && !defined(USE_ROCM)
-    std::unordered_set<std::string_view> ort_specific_ops = {"FusedConv", "QLinearAveragePool", "QLinearGlobalAveragePool"};
-#else
-    std::unordered_set<std::string_view> ort_specific_ops = {"Resize", "FusedConv", "QLinearAveragePool", "QLinearGlobalAveragePool"};
+    std::unordered_set<std::string_view> ort_specific_ops =
+    { "FusedConv",
```

Comment:
We some contrib ops like NhwcConv and GroupNorm that are for NHWC format. Shall we add them here?

Reply:
This list is for layout specific ops in NCHW format that we could insert transposes to/from NHWC around the node when [TransformLayoutForEP](https://github.com/microsoft/onnxruntime/blob/5605c3d4540099dc701a944db42079486970a987/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L851) is used. Essentially we want the list to be the opposite - NCHW format ops that an EP can request be converted to NHWC. 

If a node is in the list and is assigned to an EP that asks for NHWC layout, we will typically convert the original node in the ONNX domain to Transpose(to NHWC) ->original operator moved to internal NHWC domain -> Transpose(to NCHW). [See here](https://github.com/microsoft/onnxruntime/blob/5605c3d4540099dc701a944db42079486970a987/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L913-L918). 

The EP takes the new node in the NHWC domain and we push the Transpose as far out as possible. In the perfect case that results in all the nodes the EP has asked for being in NHWC layout. 

NhwcConv seems like it would be what the EP would use for the 'original operator moved to internal NHWC domain'.
 
Not sure about GroupNorm though. Does that take NCHW or NHWC input (or both maybe)?

Reply:
Added comment to code to explain the usage a bit more.